README
This software is copyright under GPL.  See near end file for details...

WHAT IS THIS?

Drobo-utils is a set of linux tools to query and manage Data Robotics
Drobo storage systems.   If you fire up droboview, it should look pretty
familiar to those who have seen the dashboard on other operating systems. 
Droboview is built on a little programmer interface which can be installed 
on the system and used by other applications as well.

For experienced Linux hands, there is a command line interface, drobom,
which offers the same functionality as droboview.


COMPONENTS:

  droboview   - Graphical User interface to see how the Drobo is doing.
                this is a bit like the ´dashboard´ app for other OS´s.
  drobom      - CLI script, accesses Drobo.py with no GUI stuff.
  DroboGUI.py - implements GUI. uses Drobo.py for Drobo related stuff.
  Drobo.py   -- overall Drobo io manager, provides API & "python CLI"
                uses DroboDMP.c to do raw io.
  DroboDMP.c -- python module to perform the detailed drobo ioctls.
                only platform specific stuff is in here...
  setup.py --   python version of a Makefile

REQUIREMENTS:

drobo-utils was developed on pre-release version of Kubuntu Hardy Heron.
Any similarly recent distro ought to do.

To get drobo-utils running, you need packages something like (these are
ubuntu packages, names may vary on other distros):

essential:
  python      -- interpreter for python language
  python-qt4  -- python-qt4... This is actually a new version only in 
                              the newest distros.
  libsgutils1-dev -- ioctl support.
  python-dev      -- to build DroboDMP extension.

options:
  gtksudo or kdesudo - searches on startup for one or the other.
           if neither are around, then get a graphical sudo of you choice
	   and add it to the search list at the end of DroboGUI.__init__ 
  parted   for >= 2TB file systems, need GPT support.
           just use fdisk  for smaller stuff.

INSTALLING pre-requisites.:  
On ubuntu, it would typically look like so:

	open a shell window. they enter the following package installation commands:

	% sudo aptitude install python-qt4 libsgutils1-dev libsgutils1 python-dev  
	% sudo aptitude install parted kdesudo

On redhat/fedora distros, it would more likely be 'yum' instead of 'aptitude' and
some of the package names will change.



INSTALL:
Assuming you have all of the above parts, you should be able to just do:

	python setup.py build

which on my system looks like this:
alu% python ./setup.py build
running build
running build_py
creating build
creating build/lib.linux-i686-2.5
copying Drobo.py -> build/lib.linux-i686-2.5
copying DroboGUI.py -> build/lib.linux-i686-2.5
running build_ext
building 'DroboDMP' extension
creating build/temp.linux-i686-2.5
gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include/python2.5 -c DroboDMP.c -o build/temp.linux-i686-2.5/DroboDMP.o
gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions build/temp.linux-i686-2.5/DroboDMP.o -o build/lib.linux-i686-2.5/DroboDMP.so
running build_scripts
creating build/scripts-2.5
copying and adjusting drobom -> build/scripts-2.5
copying and adjusting droboview -> build/scripts-2.5
changing mode of build/scripts-2.5/drobom from 644 to 755
changing mode of build/scripts-2.5/droboview from 644 to 755
alu%   

It should create a build/lib<somthing> sub-directory.  It will also
compile DroboDMP.c a put the result in a shared object object library 
in that directory.  If all has gone well, you can kick the tires 
by starting it up with: 
         
        ./drobom status 

see if something sensible happens... on my system with a drobo
the following happens:

% ./drobom status
/dev/sdd 100% full - ['Red alert', 'Bad disk', 'No redundancy']
%
very scary, but my drobo is in bad shape right now... you should just get []
as a status, which means there is nothing wrong.

To get all kinds of information on your drobo, try './drobom dump'
Once that is working, and assuming you have python-qt4 installed, try:

	./droboview

which should start a GUI for each drobo attached to your machine, that
you have permission to access (depends on the setup, usually USB devices 
on desktops are accessible to users, so you can see them.  Servers might be setup 
haven't worked out how to install it correctly yet.


Setup Drobo with Linux:

There is no functionality in the UI´s or API to partition or build
file systems on the Drobo.  Just use the system tools...

Drobos with firmware 1.1.1 or later work fine under linux with ext3.
You can, of course set up an NTFS or HPS+ or FAT32 if you really want,
but it seems actively counter-intuitive on Linux.  Have not tested
HPS, but ntfs-3g worked fine initially.  However, unless you are
going to physically move the disk to between systems, the native (ext3) 
format has many advantages.  The ´coffee is hot´ disclaimer is 
necessary at this point:

WARNING: THE FOLLOWING 4 LINES WILL ERASE ALL DATA ON YOUR DROBO!
WARNING: NO, IT WILL NOT ASK ANY QUESTIONS!
WARNING: ASK YOURSELF, before you start: ARE YOU SURE? 
WARNING: AFTER THE SECOND LINE, YOU ARE TOAST.
WARNING: BEST TO BACKUP YOUR DATA BEFOREHAND...

Here is what you have to type:
parted -i /dev/sdd 
mklabel gpt
mkpart ext2 0 100%
quit

The above sets up the drobo as one big partition, with a label that says
it ought to contain an ext2 file system.  If you want an NTFS file system,
then write ´ntfs´ in place of ext2.  The next step is to add the file
system into the partition.  while the above step was instantaneous, the step 
below takes a while, just have a little patience, it´ll be fine.

From the Doboshare forums, building a file system:

mke2fs –j –i 262144 –L Drobo01 -m 0 –O sparse_super,^resize_inode  /dev/sdd1

(If you want an ntfs file system, then mkntfs -f -L Drobo01 /dev/sdd1 
ought to work too... )
On my system the process looked like this:

-------------------
root@alu:~# parted -i /dev/sdd
GNU Parted 1.7.1
Using /dev/sdd
Welcome to GNU Parted! Type 'help' to view a list of commands.
(parted) mklabel gpt
(parted) mkpart ext2 0 100%
(parted) quit
root@alu:~# fdisk /dev/sdd
GNU Fdisk 1.0
Copyright (C) 1998 - 2006 Free Software Foundation, Inc.
This program is free software, covered by the GNU General Public License.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

Using /dev/sdd
Command (m for help): p

Disk /dev/sdd: 2199 GB, 2199020382720 bytes
255 heads, 63 sectors/track, 267349 cylinders
Units = cylinders of 16065 * 512 = 8225280 bytes

   Device Boot      Start         End      Blocks   Id  System
/dev/sdd1               1      267350  2147488843   83  Linux
Command (m for help): q
root@alu:~# mke2fs -j -i 262144 -L Drobo01 -m 0 -O sparse_super,^resize_inode /dev/sdd1
mke2fs 1.40.8 (13-Mar-2008)
Filesystem label=Drobo01
OS type: Linux
Block size=4096 (log=2)
Fragment size=4096 (log=2)
8388608 inodes, 536870886 blocks
0 blocks (0.00%) reserved for the super user
First data block=0
16384 block groups
32768 blocks per group, 32768 fragments per group
512 inodes per group
Superblock backups stored on blocks:
        32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208,
        4096000, 7962624, 11239424, 20480000, 23887872, 71663616, 78675968,
        102400000, 214990848, 512000000

Writing inode tables: done
Creating journal (32768 blocks): done
Writing superblocks and filesystem accounting information: done

This filesystem will be automatically checked every 26 mounts or
180 days, whichever comes first.  Use tune2fs -c or -i to override.
root@alu:~#
root@alu:~# mount /dev/sdd1 /mnt

-------------------

NOTE:

This is completely untested with multiple LUNS.  Best to make LUN
large enough to span all your space.

ON LUNSIZES >= 2TB:
 -- On older distributions, there are a couple of gotchas related to 
    linux tools which aren't 2TB ready...  to exceed 2 TB, you need to:
    	-- use GPT partitions, which aren´t supported by older fdisk
	   versions.  Tools based on libparted work fine, mostly.
    
        -- gparted fails, and seems to have a 2 TB limit on devices.
           (bug #524948 reported to bugzilla.gnome.org) It's just the GUI, 
           as libparted is fine, and other tools based on it
           still work. 




Caveats:
   droboview isn't suited to run continuously for long periods, 
   as it has a memory leak...  total foot print starts out at 32M
   with a 14 MB resident set size, of which 10 MB are shared, so only 
   about 4M of real memory consumed.   but the RSS grows at about 
   2MB/hour.

   best to restart it daily, or use it when necessary, but not leave it
   on for days.

   sadly, firmware update isn´t in there yet.

Getting a Snapshot:

   svn co https://drobo-utils.svn.sourceforge.net/svnroot/drobo-utils/trunk


   for commit access, you need a sourceforge user account.


Revision date: 2008/04/14

copyright:

Drobo Utils Copyright (C) 2008  Peter Silva (Peter.A.Silva@gmail.com)
Drobo Utils comes with ABSOLUTELY NO WARRANTY; For details type see the file
named COPYING in the root of the source directory tree.
Installation:


CHANGES
0.5.0
  * replaced DroboDMP C extension by DroboIOctl python class.  no need
    for architecture dependent packages. this also means you can just 
    try it out straight after download, no need to install anything.
    much easier.
  * device detection completely re-written.  It works better now.
  * Now displays mount points where it makes sense to do so.
  * status command output better/different. shows all devices and mount points
    for each drobo.
  * full Multiple LUN support.    should do what is expected.
  * full multiple Drobo support.  should do what is expected.
  * tested with multiple devices, now generally works in all cases.
  * added modular info printing with csv list of outputs.
  * firmware 1.3.0, has enabled formerly broken functionality which
    name makes the sync command rename every drobo to 'hi there'.
    fixed.
  * added 'name' command, to name Drobo. (functionality of Sync)
  * corrected detection logic, which depended on formerly immutable name. 
    (affects fw >= 1.3.0) now use same mechanisms as sg_scan from sg3_utils.

0.4.0 Release: 2008/12/28
  * IMPORTANT: firmware 1.3.0 support added.  Earlier drobo-utils 
     versions will not recognize a drobo running 1.3.0 or later.
  * CLI is more normal...  use getopts (got rid of device as fixed place arg.)
     now allows command to be an option as well.  
  * Changed DEBUG from a flag requiring code modification, to a bitmap
     honored as a command-line argument (--verbose)
  * misc. improvements in documentation.
  * droboview now only launches for first drobo found in list, and stays foreground
     always.
  * drobom list returns a list that is easier to parse (suitable for use in backticks.)
  * droboview is now just a stub for drobom view. reduced overhead.
  * now reports VERSION id in usage.  There is a -V option too.


0.3.3 release: 2008/11/05
  * lunsize display bug fixed.
  * Chris's man pages & help improvements.
  * more messaging fixups for when you don't invoke as root.
  * fix for ubuntu Intrepid deciding all disks are Drobos. !!
  * fixed all the firmware info parsing issues.
  * firmware load issues should all be gone.
  * Brad's fix for ddiag.c, done properly this time (my bad!)

0.3.2 release 2008/10/25

  * for v1, still don't understand index.txt file.
  * after testing, some more fixes, v2 downloads now work.
  * fixes received from Brad Guillory for v2/tdz firmware downloading.
     they don't do any harm afaict, don't have a v2 to test with.
  * added fwload directive to drobom.
  * added root user check to give a bigger hint.

0.3.1 release 2008/10/01

  * OK I know checking for firmware updates doesn't work
     right now, but I need to find out why.  code seems correct.
     there is a bugfix for getting rid of the 'licensed' part.

  * fixes for firmware validation:
       * header CRC on 32bit was broken on 64bit.
       * 32 vs. 64 bit kludge needed.

0.3.0 release 2008/09/04
  * added ability to format Drobo
  * added ability to set lun sizes
  * bugfix for firmware:
        * CRC32 on 32bit intel. (use signed, instead of unsigned.)
        * fixed header CRC not validating.
  * added ability to set time on Drobo.
  * added simulation mode to Drobo.py, development aid.
  * some work done towards lunsize setting, but incomplete.
  * some work towards indicating relay out progress, not tested...

0.2.2 Released 2008/08/08
  * firmware uploads work with .tdz as well
  * tested with older firmwares, contains a few fixes to improve compatibility.
     works well enought upgrade old firmwares.

0.2.1 Released: 2008/08/06 
  * uploaded to drobospace & sourceforge.
  * firmware upload works for .tdf's 
     
0.1.1 Initial version uploaded  april 2008
  * to drobospace.com and sf.net
  * confirmed to work with firmware 1.1.1
  * shows status, disk slots filled, model, etc..
  * shows fw version loaded.
  * blink & standby commands work.


DroboGUI.py
#!/usr/bin/python

#Drobo Utils Copyright (C) 2008  Peter Silva (Peter.A.Silva@gmail.com)
#Drobo Utils comes with ABSOLUTELY NO WARRANTY; For details type see the file
#named COPYING in the root of the source directory tree.
#

#-------------------------------------------------------------------
# debugging stuff: begin
#-------------------------------------------------------------------

#import pdb


#------------------------------------------------------------------
# debugging stuff: end
#-------------------------------------------------------------------
import os,sys,math
from PyQt4 import QtGui
from PyQt4 import QtCore
import socket
import Drobo
import subprocess
import commands
import string

def _toGB(num):
  g = num*1.0/(1000*1000*1000)
  return "%6.1f" % g

def _toTiB(num):
  """
  convert input number to computerish Terabytes.... (ok... TibiBytes blech...)

  STATUS: works bizarrely...
  bug in python 2.5? - a number which is 2 TB -1, + any number < 5000 ends up smaller than 2TiB.  so at first I just added 5000 to get the correct answer. did a manual binary search, and even 4096 doesn't work... but when I change OS's, even that didn't workreliably...  so now I just round it at each division...
  """
  #print num, num/1024, num/(1024*1024), num/(1024*1024*1024), 
  g = round(round(round(round(num/1024)/1024)/1024)/1024)
  return int(g)
 


def _setDiskLabel(model,capacity): 
    if (capacity == ''):
      label = model
    else:
      if ( capacity > 0):
           label = model.rstrip() + '   ' + _toGB(capacity) + 'GB '
      else:
	   label = 'empty'
    return label

partitioner=""

def _runPartitioner():
   """ invoke existing partitioning program...
   """
   print "partitioner = ", partitioner
   subprocess.Popen( partitioner, shell=True)

class DroboAbout(QtGui.QWidget):
   def __init__(self, parent=None):
        QtGui.QWidget.__init__(self, parent)
        self.main = QtGui.QLabel("""
  drobo-utils: software to manage a Drobo storage unit from Data Robotics International Corp.
  Winner of the Data Robotics Bounty 2008 for a linux dashboard!  
  Home page:  http://drobo-utils.sourceforge.net

  Version: """ + Drobo.VERSION + """ 
  Copyright 2008 Peter Silva ( Peter.A.Silva@gmail.com )
  license: General Public License (GPL) v3

  with contributions from: Chris Atlee (chris@atlee.ca), Brad Guillory, <withheld@spammenot.norge> 
  and inspiration from: Joe Krahn

  and thanks for DRI for putting up the Bounty!
        """ , self)
        self.setMinimumSize(240, 240)
        self.quit = QtGui.QPushButton('Quit',self)
        self.quit.setCheckable(False)
        self.quit.setMinimumWidth(200)
        #self.quit.move(-30,-100)
        self.connect(self.quit, QtCore.SIGNAL('clicked()'), 
                self.hide)
 


class DroboGUI(QtGui.QMainWindow):
    """ GUI for a single Drobo, start one for each drobo.
    """

    def __doButton(self,x,y,parent):
	""" create a disk display "button" ...
        """
        button = QtGui.QPushButton('uninitialized - 0000GB', parent)
        button.setCheckable(False)
        button.setMinimumWidth(200)
        button.setStyleSheet( "QWidget { background-color: white }" )
        button.move(x,y)
        #self.connect(button, QtCore.SIGNAL('focusInEvent()'), 
	#	self.__StatusBar_space)
        return button

    def __Blink(self):
        self.drobo.Blink()

    def __StatusBar_space(self):
        c=self.drobo.GetSubPageCapacity()
        self.statusmsg = 'used: ' + _toGB(c[1]) + ' free: ' + _toGB(c[0]) + ' Total: ' + _toGB(c[2]) + ' GB, update# ' + str(self.updates)
	#print self.statusmsg
	
    def __updateLEDs(self):
        """ update LEDS (implement flashing.)
        """

        self.updates = self.updates + 1 
        i=0
        while ( i < 4 ):
          c=self.s[i][3]
          if (len(c) == 2 ):  
             colour = c[ self.updates % 2] 
          else:
	     colour = c

          self.Device.slot[i][1].setStyleSheet(
                "QWidget { background-color: %s }" % colour )
	  i=i+1


    def __updatewithQueryStatus(self):
        """ query device to update information
        """
        try:
           fwv=self.drobo.GetSubPageFirmware()
        except:
           self.statusBar().showMessage( 'bad poll: %d.. need to restart' % self.updates )
           return

        settings=self.drobo.GetSubPageSettings()
        self.Device.id.setText(  self.drobo.GetCharDev() + ' firmware: ' + fwv[7] )

        self.Device.id.setToolTip(
           "Firmware build: " + str(fwv[0]) + '.' + str(fwv[1]) + '.' + str(fwv[2]) + "\n Features: " + string.join(fwv[8],"\n") )
        
        self.s=self.drobo.GetSubPageSlotInfo()

        luninfo=self.drobo.GetSubPageLUNs()

        luntooltip="luns, count: " + str(len(luninfo)) + "\n"
        for l in luninfo:
	   luntooltip = luntooltip + "lun id: " + str(l[0]) + " used: " +  \
                   _toGB(l[2]) + " total: " + _toGB(l[1]) 
           if 'SUPPORTS_NEW_LUNINFO2' in self.drobo.features :
              luntooltip = luntooltip + " scheme: " + l[3] + " type: " + str(l[4]) 
           luntooltip = luntooltip + "\n"

        i=0
        while ( i < 4 ):
          self.Device.slot[i][0].setText(_setDiskLabel(self.s[i][5],self.s[i][1]))
          self.Device.slot[i][0].setToolTip(luntooltip)
	  i=i+1

        c=self.drobo.GetSubPageConfig()
        self.Format.lunsize =  _toTiB(c[2])

        c=self.drobo.GetSubPageCapacity()
        if c[2] > 0:
           self.Device.fullbar.setValue( c[1]*100/c[2] )
           self.Device.fullbar.setToolTip( 
	    "used: " + _toGB(c[1]) + ' free: ' + _toGB(c[0]) + ' Total: ' + _toGB(c[2]) + ' GB, update# ' + str(self.updates) )
	#print self.statusmsg
        #self.__StatusBar_space()
        self.statusBar().showMessage( self.statusmsg )
        ss = self.drobo.GetSubPageStatus()
        self.statusmsg = 'Status: ' + str(ss[0]) + ' update: ' + str(self.updates)

        if self.Format.inProgress and ( self.fmt_process.poll() != None) :
               # reset to normal state...
               print 'it took: %d updates to run' % (self.updates - self.Format.startupdate )
               self.Format.inProgress=0
               normal = self.Tools.Updatebutton.palette().color( QtGui.QPalette.Button )
               self.Format.Formatbutton.palette().setColor( QtGui.QPalette.Button, QtCore.Qt.blue )
               self.Format.ext3.setChecked(0)
               self.Format.ntfs.setChecked(0)
               self.Format.msdos.setChecked(0)
               self.Format.Formatbutton.setText('Format Done!')
               self.Format.connect(self.Format.Formatbutton, QtCore.SIGNAL('clicked()'),
                       self.FormatLUN)
        #pdb.set_trace()

    def __updateStatus(self):
        self.__updateLEDs() # only update display...

        # try not to poll the device too often, so only every 'n' screen updates
        if (self.updates % 5 == 0 ): # query device for new info...
	    self.__updatewithQueryStatus()


    def __initDeviceTab(self):

        self.Device = QtGui.QWidget()
        self.Device.setObjectName("Device")

        # Create Device tab...
        x=10
        y=30 
        yby=35

        self.Device.id =self.__doButton(x,y,self.Device)
        self.Device.id.setMinimumWidth(220)
 
        self.Device.slot = [ [ '', '' ], [ '', '' ], [ '', '' ], [ '', '' ] ]
        y=y+yby

        i=0
        while ( i < 4 ):
          self.Device.slot[i][0] =self.__doButton(x,y,self.Device)
          y=y+yby
          i=i+1

        x=x+210
        y=30+yby
        i=0
        while ( i < 4 ):
          self.Device.slot[i][1] = QtGui.QWidget(self.Device)
          self.Device.slot[i][1].setGeometry(x, y, 10, 30)
          y=y+yby
          i=i+1

        x=10
        self.Device.fullbar = QtGui.QProgressBar(self.Device)
        self.Device.fullbar.setGeometry(x, y, 230, 25)
        self.connect(self.Device.fullbar, QtCore.SIGNAL('focusInEvent()'), 
		self.__StatusBar_space)

        self.tab.addTab(self.Device, "Device")

    def ReallyFormatLUN(self):

       print 'Really formatting...'
       self.Format.disconnect(self.Format.Formatbutton, QtCore.SIGNAL('clicked()'),
                self.ReallyFormatLUN)
       if self.Format.fstype == 'none': # changing LUN size
            self.Format.Formatbutton.setText('Done. WAIT 5 min. restart dashboard!')
            self.drobo.SetLunSize(self.Format.lunszlcd.value())
       else:
            self.Format.Formatbutton.setText('Format in progress, WAIT!')
            format_script = self.drobo.format_script(self.Format.fstype)
            self.Format.startupdate = self.updates
            self.Format.inProgress = 1;
            #self.fmt_process = subprocess.Popen( format_script, bufsize=0, stdout=subprocess.PIPE )
            self.fmt_process = subprocess.Popen( format_script )
            p = self.Format.Formatbutton.palette()
            p.setColor( QtGui.QPalette.Button, QtCore.Qt.red )




    def FormatLUN(self):
       print 'Clicked format...'
       if self.Format.ntfs.isChecked():
            fstype='ntfs'
       elif self.Format.msdos.isChecked():
            fstype='FAT32'
       elif self.Format.ext3.isChecked():
            fstype='ext3'
       else:
            fstype='none'
            self.Format.Formatbutton.palette().setColor( QtGui.QPalette.Button, QtCore.Qt.yellow )
            self.Format.Formatbutton.setText( \
                  "Last Chance, Resize to %d TiB" % self.Format.lunszlcd.value() )

       if fstype != 'none':
          self.Format.Formatbutton.palette().setColor( QtGui.QPalette.Button, QtCore.Qt.yellow )
          self.Format.Formatbutton.setText( "Last Chance, Format %s ?" % fstype)

       self.Format.fstype=fstype
       self.Format.disconnect(self.Format.Formatbutton, QtCore.SIGNAL('clicked()'),
                self.FormatLUN)
       self.Format.connect(self.Format.Formatbutton, QtCore.SIGNAL('clicked()'),
                self.ReallyFormatLUN)

    def __adjustlunsize(self,sz):

        newsize=2**sz
        self.Format.lunszlcd.display(newsize)
        if ( self.Format.lunsize != newsize ):
            self.Format.fstype= 'none'
            self.Format.ext3.setChecked(False)
            self.Format.ext3.setCheckable(False)
            self.Format.msdos.setChecked(False)
            self.Format.msdos.setCheckable(False)
            self.Format.ntfs.setChecked(False)
            self.Format.ntfs.setCheckable(False)
            self.Format.Formatbutton.setText( "Set LUN size to %d TiB" % newsize)
        else:
            self.Format.ext3.setCheckable(True)
            self.Format.msdos.setCheckable(True)
            self.Format.ntfs.setCheckable(True)
            self.Format.Formatbutton.setText( "Format (Erases All Data!)         " )


    def __initFormatTab(self):
        """
          known issues:
           -- formats only using GPT labels... should be mbr for FAT32.
           -- should refuse to set lunsize > 2 TB for FAT32.
           -- should refuse to set lunsize > 8 TB for ext3.

        """

	self.Format = QtGui.QWidget()
        self.Format.setObjectName("Format")
        xo=20
        w=160
        h=16
        s=10

        x=10
        y=20
        self.Format.header = QtGui.QLabel("WARNING: Format erases whole Drobo", self.Format)
        self.Format.header.setStyleSheet( "QWidget { color: red }" )
        self.Format.header.move(x,y)

        x=xo
        y=y+h+s
        self.Format.lunsztitle = QtGui.QLabel("Maximum LUN size:", self.Format)
        self.Format.lunsztitle.move(x,y)

        self.Format.lunszlcd = QtGui.QLCDNumber(2, self.Format)

        x=x+120
        self.Format.lunszlcd.move(x,y)

 
        x=xo
        y=y+h+s

        self.Format.horizontalSlider = QtGui.QSlider(self.Format)
        self.Format.horizontalSlider.setGeometry(QtCore.QRect(x,y,w,h))
        self.Format.horizontalSlider.setMaximum(4)
        self.Format.horizontalSlider.setMinimum(0)
        self.Format.horizontalSlider.setSingleStep(1)
        self.Format.horizontalSlider.setPageStep(2)
        self.Format.horizontalSlider.setOrientation(QtCore.Qt.Horizontal)
        self.Format.horizontalSlider.setTickPosition(QtGui.QSlider.TicksBelow)
        self.Format.horizontalSlider.setTickInterval(1)
        self.Format.horizontalSlider.setObjectName("horizontalSlider")

        self.Format.horizontalSlider.setProperty("value",QtCore.QVariant(2))
        c=self.drobo.GetSubPageConfig()
        self.Format.lunsize =  _toTiB(c[2])
        self.Format.lunszlcd.display( self.Format.lunsize )
        self.Format.horizontalSlider.setValue( int(math.log(self.Format.lunsize,2)) )

        self.Format.connect(self.Format.horizontalSlider, QtCore.SIGNAL('valueChanged(int)'),
                self.__adjustlunsize)
        #self.Format.connect(self.Format.horizontalSlider, QtCore.SIGNAL('valueChanged(int)'),
        #        self.Format.lunszlcd.display)

        y=y+h+s
        self.Format.ext3 = QtGui.QRadioButton("Ext3 (journalled ext2)", self.Format)
        mkfs = commands.getoutput("which mke2fs")
        if ( mkfs == "" ): 
            self.Format.ext3.setCheckable(0)
            self.Format.ext3.setStyleSheet( "QWidget { color: gray }" )
            self.Format.ext3.setText('Ext3 disabled (missing mke2fs)')

        self.Format.ext3.move(x,y)

        y=y+h+s
        self.Format.msdos = QtGui.QRadioButton("FAT32 MS - Disk Operating System", self.Format)
        #self.Format.msdos.setStyleSheet( "QWidget { color: gray }" )
        mkfs = commands.getoutput("which mkdosfs")
        if ( mkfs == "" ): 
            self.Format.msdos.setCheckable(0)
            self.Format.msdos.setStyleSheet( "QWidget { color: gray }" )
            self.Format.msdos.setText('FAT32 disabled (missing mkdosfs)')

        self.Format.msdos.move(x,y)

        y=y+h+s
        self.Format.ntfs = QtGui.QRadioButton("NTFS -- Windows NT/XP/Vista", self.Format)
        #self.Format.ntfs.setStyleSheet( "QWidget { color: gray }" )
        self.Format.ntfs.move(x,y)

        mkfs = commands.getoutput("which mkntfs")
        if ( mkfs == "" ): 
            self.Format.ntfs.setCheckable(0)
            self.Format.ntfs.setStyleSheet( "QWidget { color: gray }" )
            self.Format.ntfs.setText('NTFS disabled (missing mkntfs)')

        y=y+h+s

        self.Format.Formatbutton = QtGui.QPushButton('Format (Erases All Data!)       ', self.Format)
        #self.Format.Formatbutton.setStyleSheet( "QWidget { color: gray }" )
        self.Format.Formatbutton.move(x,y)

        self.tab.addTab(self.Format, "Format")

        self.Format.connect(self.Format.Formatbutton, QtCore.SIGNAL('clicked()'),
                self.FormatLUN)

        # progress flag...
        self.Format.inProgress = 0


    def upgrade(self):
        self.disconnect(self.Tools.Updatebutton, QtCore.SIGNAL('clicked()'), self.upgrade)
        self.connect(self.Tools.Updatebutton, QtCore.SIGNAL('clicked()'), self.checkup)

        if self.drobo.updateFirmwareRepository():
             self.drobo.writeFirmware( self.Tools.progress.setValue )
        self.Tools.comment.setText( "Written! Reboot Drobo to activate.")

    def checkup(self):
        self.drobo.Sync() # convenient side effect:  make the host and drobo clocks agree...
        (fwarch, fwversion, hwlevel, fwpath ) = self.drobo.PickLatestFirmware()
        print "checkup: this Drobo is a %s hw rev: %s, and needs: %s" % ( fwarch, hwlevel, fwversion )
        if fwpath != '' :
            self.Tools.Updatebutton.setText( "Upgrade" )
            self.disconnect(self.Tools.Updatebutton, QtCore.SIGNAL('clicked()'), self.checkup)
            self.connect(self.Tools.Updatebutton, QtCore.SIGNAL('clicked()'), 
                self.upgrade)
            self.Tools.comment.setText( "Press 'Upgrade' upgrade to %s" % ( fwversion ))
        else:
            self.Tools.comment.setText( "No update available!" )
     
        
    def __diags(self):
        self.drobo.Sync() # convenient side effect  make the host and drobo clocks agree...
        fname = self.drobo.dumpDiagnostics()
        self.Tools.comment.setText( fname )

    def __initToolTab(self):

	self.Tools = QtGui.QWidget()
        self.Tools.setObjectName("Tools")
        xo=20
        x=xo
        y=30
        w=160
        h=16
        s=10

        #
        # Set the tool colors to grey, to indicate non-functional...
        #
        self.Tools.Standbybutton = QtGui.QPushButton('Shutdown', self.Tools)
        self.Tools.Standbybutton.setCheckable(False)
        self.Tools.Standbybutton.move(x,y)

        self.connect(self.Tools.Standbybutton, QtCore.SIGNAL('clicked()'), 
                self.drobo.Standby)

	w=self.Tools.Standbybutton.width()
        x=x+w+s

        self.Tools.Blinkybutton = QtGui.QPushButton('Blink Lights', self.Tools)
        self.Tools.Blinkybutton.setCheckable(False)
        self.Tools.Blinkybutton.move(x,y)

        self.connect(self.Tools.Blinkybutton, QtCore.SIGNAL('clicked()'), 
                self.drobo.Blink)

        h=self.Tools.Standbybutton.height()
        x=xo
        y=y+h+s
        self.Tools.Renamebutton = QtGui.QPushButton('Rename', self.Tools)
        self.Tools.Renamebutton.setStyleSheet( "QWidget { color: gray }" )
        self.Tools.Renamebutton.setCheckable(False)
        self.Tools.Renamebutton.move(x,y)
        
        x=x+w+s
        self.Tools.Updatebutton = QtGui.QPushButton('Update', self.Tools)
        self.Tools.Updatebutton.move(x,y)

        self.connect(self.Tools.Updatebutton, QtCore.SIGNAL('clicked()'), self.checkup)


        x=xo
        y=y+h+s
        Registerbutton = QtGui.QPushButton('Register', self.Tools)
        Registerbutton.setStyleSheet( "QWidget { color: gray }" )
        Registerbutton.setCheckable(False)
        Registerbutton.move(x,y)

        x=x+w+s
        Diagbutton = QtGui.QPushButton('Diagnostics', self.Tools)
        Diagbutton.setCheckable(False)
        Diagbutton.move(x,y)

        self.connect(Diagbutton, QtCore.SIGNAL('clicked()'), 
                self.__diags)

        # next button...
        x=xo
        y=y+h+s

        self.Tools.progress = QtGui.QProgressBar(self.Tools)
        self.Tools.progress.setGeometry(x, y, 180, 25)
        self.Tools.progress.setValue( 0 )

        x=xo
        y=y+h+s

        self.Tools.comment = QtGui.QLabel("Press 'Update' to look for updates", self.Tools)
        self.Tools.comment.move(x,y)

        normal = self.Tools.Updatebutton.palette().color( QtGui.QPalette.Button )

        self.tab.addTab(self.Tools, "Tools")


    def __init__(self, d, parent=None):
        QtGui.QMainWindow.__init__(self)
        #QtGui.QWidget.__init__(self, parent)

	global partitioner

        self.drobo = d
        self.updates = 0

        self.statusmsg='Ready'
        self.color = QtGui.QColor(0, 0, 0) 
        #QtGui.QApplication.setStyle(QtGui.QStyleFactory.create('cleanlooks'))
        QtGui.QApplication.setStyle(QtGui.QStyleFactory.create('MacStyle'))
        #self.setStyleSheet( "QWidget { border-color: white }" )
        #self.setStyleSheet( "QWidget { background-color: black }" )
        #self.setStyleSheet( "QWidget { color: white }" )

        #self.setGeometry(300, 100, 350, 260)
        self.setMinimumSize(240, 310)
        self.setWindowTitle('DroboView')
        self.setWindowIcon(QtGui.QIcon(':Drobo-Front-0000.gif'))

        exit = QtGui.QAction(QtGui.QIcon('icons/exit.png'), 'Exit', self)
        exit.setShortcut('Ctrl+Q')
        exit.setStatusTip('Exit application')
        self.connect(exit, QtCore.SIGNAL('triggered()'), QtCore.SLOT('close()'))

        menubar = self.menuBar()
        file = menubar.addMenu('&File')
        file.addAction(exit)

        #parted = QtGui.QAction(QtGui.QIcon('icons/exit.png'), 'Manage Space', self)
        #parted.setShortcut('Ctrl+S')
        #parted.setStatusTip('Start up gparted Space Manager')
        #self.connect(parted, QtCore.SIGNAL('triggered()'), _runPartitioner)
        #tools = menubar.addMenu('&Tools')
        #tools.addAction(parted)

        about = QtGui.QAction(QtGui.QIcon('icons/exit.png'), 'About DroboView', self)
        self.aboutdialog = DroboAbout()
        help = menubar.addMenu('&Help')
        help.addAction(about)
        self.connect(about, QtCore.SIGNAL('triggered()'), self.aboutdialog.show)

        self.tab = QtGui.QTabWidget(self)
	self.tab.setGeometry(QtCore.QRect(0,27,240,260))
        self.tab.setObjectName("tabs")
        
	self.__initDeviceTab()
	self.__initToolTab()
	self.__initFormatTab()

	self.__updatewithQueryStatus()
        self.updateTimer = QtCore.QTimer(self)
        self.connect(self.updateTimer, QtCore.SIGNAL("timeout()"),
		self.__updateStatus )
        self.updateTimer.setInterval(1000)
        self.updateTimer.start()

        partitioner= " gparted"

